docs/DdnRelocatableShaderElf.md

```diff
@@ -0,0 +1,205 @@
+
+Fragment shaders will require a prologue to export its outputs.  The unlinked fragment
+shader will have the standard name for the fragment shader.  Instead of
+exporting the outputs, they are left in registers at the end of the shader,
+where there is no `s_endpgm`. Extra information is add to the PAL metadata in
+the `.colorExport` table.  During linking this table and the pipeline state are
+used to generate all of the color export instructions and they are appended to
+the end of the unlinked shader. See the `ColorExportShader` class in
+[GlueShader.cpp](https://github.com/GPUOpen-Drivers/llpc/blob/dev/lgc/elfLinker/GlueShader.cpp).
+for the generation of the epilogue.
+To see how the unlinked shader handles
+color export see the `LowerFragColorExport` pass in
+[FragColorExport.cpp](https://github.com/GPUOpen-Drivers/llpc/blob/dev/lgc/patch/FragColorExport.cpp).
+
+These epilogues and prologues mean that unlinked shader do not have to depend on
+the format of the vertex input attributes and color exports.
+
+## User data nodes
+
+The user data nodes are handled differently in unlinked shaders. The
+differences are best described with examples.  Suppose you want to load the
+descriptor for descriptor set 2 at binding 1.
+
+1. Suppose the descriptor set is not a top level user data node.  Then the
+   standard compile will add a parameter to the shader for the top level user data node
+   that contains descriptor set 2.  The corresponding `SPI_SHADER_USER_DATA`
+   entry in the PAL meta data will contain the offset of that top level node.
+   An unlinked shader may not have that offset available.  Instead the PAL metadata entry
+   will contain the number of the descriptor set that the parameter is suppose
+   to point to.  In this case, it will be 2.  In
+   `PalMetadata::fixUpRegisters` in
+   [PalMetadata.cpp](https://github.com/GPUOpen-Drivers/llpc/blob/dev/lgc/state/PalMetadata.cpp)
+   the offset is filled in.
+
+1. Suppose the descriptor set is in a top level user data node.  Then the
+   standard compile could get the descritor directly from an input register.  When
+   building an unlinked shader, this is not possible.  There are not enough
+   input register to have a 
+   separate input register for every descriptor that is potentially in a top
+   level node.  So for unlinked shaders, this is handled by spilling the user
+   data node, and indirectly loading the descriptor from the spill table.  This
+   is done in `DescBuilder::CreateLoadBufferDesc` in 
+   [DescBuilder.cpp](https://github.com/GPUOpen-Drivers/llpc/blob/dev/lgc/builder/DescBuilder.cpp).
+
+1. The final difficulty is that an unlinked shader cannot know the difference
+   between the two previous cases.  For cases where the descriptor is possibly
+   in a top level user data node, the compiler will generate code that will
+   select either the spill table or the entry for that descriptor set based on a
+   relocation. See `DescBuilder::getDescPtr` in
```

Comment:
Consider adding a GCN snippet with the code pattern mentioned

Reply:
I've added a new test that I point to.